File: Raven.Client.Lightweight/Linq/DynamicAggregationQuery.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using System.Threading.Tasks;
using Raven.Abstractions.Data;
using Raven.Abstractions.Extensions;

namespace Raven.Client.Linq
{

	public class DynamicAggregationQuery<T>
	{
		private readonly IQueryable<T> queryable;
		private readonly List<AggregationQuery<T>> facets;
        private readonly Dictionary<string,string> renames = new Dictionary<string, string>();

		public DynamicAggregationQuery(IQueryable<T> queryable, Expression<Func<T, object>> path, string displayName = null)
		{
			facets = new List<AggregationQuery<T>>();
			this.queryable = queryable;
			AndAggregateOn(path, displayName);
		}


		public DynamicAggregationQuery(IQueryable<T> queryable, string path, string displayName = null)
		{
			facets = new List<AggregationQuery<T>>();
			this.queryable = queryable;
			facets.Add(new AggregationQuery<T> { Name = path, DisplayName = displayName });

		}

		public DynamicAggregationQuery<T> AndAggregateOn(Expression<Func<T, object>> path, string displayName = null)
		{
			var propertyPath = path.ToPropertyPath('_');
			if (IsNumeric(path))
			{
				var tmp = propertyPath + "_Range";
				renames[propertyPath] = tmp;
				propertyPath = tmp;
			}

			facets.Add(new AggregationQuery<T> { Name = propertyPath, DisplayName = displayName});

			return this;
		}

		public DynamicAggregationQuery<T> AndAggregateOn(string path, string displayName = null)
		{
			facets.Add(new AggregationQuery<T> { Name = path, DisplayName = displayName });

			return this;
		}

	    private bool IsNumeric(Expression<Func<T, object>> path)
	    {
	        var unaryExpression = path.Body as UnaryExpression;
	        if (unaryExpression == null)
	            return false;
	        if (unaryExpression.NodeType != ExpressionType.Convert &&
	            unaryExpression.NodeType != ExpressionType.ConvertChecked)
	            return false;
	        var type = unaryExpression.Operand.Type;
	        return type == typeof (int) ||
	               type == typeof (long) ||
	               type == typeof (short) ||
	               type == typeof (decimal) ||
	               type == typeof (double) ||
	               type == typeof (float);
	    }

	    public DynamicAggregationQuery<T> AddRanges(params Expression<Func<T, bool>>[] paths)
		{
			var last = facets.Last();
			
			last.Ranges = last.Ranges ?? new List<Expression<Func<T, bool>>>();

			foreach (var func in paths)
			{
				last.Ranges.Add(func);
			}

			return this;
		}

	    private void SetFacet(Expression<Func<T, object>> path, FacetAggregation facetAggregation)
		{
			var last = facets.Last();
			last.AggregationField = path.ToPropertyPath();
		    last.AggregationType = path.ExtractTypeFromPath().FullName;
			last.Aggregation |= facetAggregation;
		}

		public DynamicAggregationQuery<T> MaxOn(Expression<Func<T, object>> path)
		{
			SetFacet(path, FacetAggregation.Max);

			return this;
		}

		public DynamicAggregationQuery<T> MinOn(Expression<Func<T, object>> path)
		{
			SetFacet(path, FacetAggregation.Min);

			return this;
		}

		public DynamicAggregationQuery<T> SumOn(Expression<Func<T, object>> path)
		{
			SetFacet(path, FacetAggregation.Sum);

			return this;
		}

		public DynamicAggregationQuery<T> AverageOn(Expression<Func<T, object>> path)
		{
			SetFacet(path, FacetAggregation.Average);

			return this;
		}

		public DynamicAggregationQuery<T> CountOn(Expression<Func<T, object>> path)
		{
			SetFacet(path, FacetAggregation.Count);

			return this;
		}

#if !NETFX_CORE
		public FacetResults ToList()
		{
			return HandlRenames(queryable.ToFacets(AggregationQuery<T>.GetFacets(facets)));
		}

		public Lazy<FacetResults> ToListLazy()
		{
			var facetsLazy = queryable.ToFacetsLazy(AggregationQuery<T>.GetFacets(facets));
			return new Lazy<FacetResults>(() => HandlRenames(facetsLazy.Value));
		}
#endif

		public async Task<FacetResults> ToListAsync()
		{
			return HandlRenames(await queryable.ToFacetsAsync(AggregationQuery<T>.GetFacets(facets)));
		}

	    private FacetResults HandlRenames(FacetResults facetResults)
	    {
	        foreach (var rename in renames)
	        {
	            FacetResult value;
	            if (facetResults.Results.TryGetValue(rename.Value, out value) &&
	                facetResults.Results.ContainsKey(rename.Key) == false)
	                facetResults.Results[rename.Key] = value;
	        }
	        return facetResults;
	    }
	}
}

File: Raven.Client.Java/lib/src/main/java/raven/client/linq/DynamicAggregationQuery.java
package raven.client.linq;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import raven.abstractions.basic.Lazy;
import raven.abstractions.closure.Function0;
import raven.abstractions.data.FacetAggregation;
import raven.abstractions.data.FacetResult;
import raven.abstractions.data.FacetResults;
import raven.abstractions.extensions.ExpressionExtensions;

import com.mysema.query.types.Path;
import com.mysema.query.types.expr.BooleanExpression;

public class DynamicAggregationQuery<T> {
  private final IRavenQueryable<T> queryable;
  private final List<AggregationQueryDsl> facets;
  private final Map<String, String> renames = new HashMap<String, String>();

  public DynamicAggregationQuery(IRavenQueryable<T> queryable, Path<?> path) {
    this(queryable, path, null);
  }

  public DynamicAggregationQuery(IRavenQueryable<T> queryable, Path<?> path, String displayName) {
    facets = new ArrayList<>();
    this.queryable = queryable;
    andAggregateOn(path, displayName);
  }

  public DynamicAggregationQuery(IRavenQueryable<T> queryable, String path) {
    this(queryable, path, null);
  }

  public DynamicAggregationQuery(IRavenQueryable<T> queryable, String path, String displayName) {
    facets = new ArrayList<>();
    this.queryable = queryable;
    AggregationQueryDsl aggregationQueryDsl = new AggregationQueryDsl();
    aggregationQueryDsl.setName(path);
    aggregationQueryDsl.setDisplayName(displayName);
    facets.add(aggregationQueryDsl);
  }

  public DynamicAggregationQuery<T> andAggregateOn(Path<?> path) {
    return andAggregateOn(path, null);
  }

  public DynamicAggregationQuery<T> andAggregateOn(Path<?> path, String displayName) {
    String propertyPath = ExpressionExtensions.toPropertyPath(path);
    if (isNumeric(path)) {
      String tmp = propertyPath + "_Range";
      renames.put(propertyPath, tmp);
      propertyPath = tmp;
    }

    AggregationQueryDsl aggregationQuery = new AggregationQueryDsl();
    aggregationQuery.setName(propertyPath);
    aggregationQuery.setDisplayName(displayName);
    facets.add(aggregationQuery);
    return this;
  }

  public DynamicAggregationQuery<T> andAggregateOn(String path) {
    return andAggregateOn(path, null);
  }

  public DynamicAggregationQuery<T> andAggregateOn(String path, String displayName) {
    AggregationQueryDsl aggregationQuery = new AggregationQueryDsl();
    aggregationQuery.setName(path);
    aggregationQuery.setDisplayName(displayName);
    facets.add(aggregationQuery);
    return this;
  }

  private boolean isNumeric(Path<?> path) {
    Class< ? > type = path.getType();
    return int.class.equals(type)
        || long.class.equals(type)
        || short.class.equals(type)
        || double.class.equals(type)
        || float.class.equals(type);
  }

  public DynamicAggregationQuery<T> addRanges(BooleanExpression... paths) {
    AggregationQueryDsl last = facets.get(facets.size() - 1);
    if (last.getRanges() == null) {
      last.setRanges(new ArrayList<BooleanExpression>());
    }
    for (BooleanExpression expr: paths) {
      last.getRanges().add(expr);
    }
    return this;
  }

  private void setFacet(Path<?> path, FacetAggregation facetAggregation) {
    AggregationQueryDsl last = facets.get(facets.size() - 1);
    last.setAggregrationField(ExpressionExtensions.toPropertyPath(path));
    last.setAggregationType(path.getType().getName());
    last.getAggregation().add(facetAggregation);
  }

  public DynamicAggregationQuery<T> maxOn(Path<?> path) {
    setFacet(path, FacetAggregation.MAX);
    return this;
  }

  public DynamicAggregationQuery<T> minOn(Path<?> path) {
    setFacet(path, FacetAggregation.MIN);
    return this;
  }

  public DynamicAggregationQuery<T> sumOn(Path<?> path) {
    setFacet(path, FacetAggregation.SUM);
    return this;
  }

  public DynamicAggregationQuery<T> averageOn(Path<?> path) {
    setFacet(path, FacetAggregation.AVERAGE);
    return this;
  }

  public DynamicAggregationQuery<T> countOn(Path<?> path) {
    setFacet(path, FacetAggregation.COUNT);
    return this;
  }

  public FacetResults toList() {
    return handleRenames(queryable.toFacets(AggregationQueryDsl.getFacets(facets)));
  }

  public Lazy<FacetResults> toListLazy() {
    final Lazy<FacetResults> facetsLazy = queryable.toFacetsLazy(AggregationQueryDsl.getFacets(facets));
    return new Lazy<>(new Function0<FacetResults>() {

      @Override
      public FacetResults apply() {
        return handleRenames(facetsLazy.getValue());
      }
    });
  }

  private FacetResults handleRenames(FacetResults facetResults) {
    for (Map.Entry<String, String> rename: renames.entrySet()) {
      if (facetResults.getResults().containsKey(rename.getValue())
          && !facetResults.getResults().containsKey(rename.getKey())) {
        FacetResult value = facetResults.getResults().get(rename.getValue());
        facetResults.getResults().put(rename.getKey(), value);
      }
    }
    return facetResults;
  }

}
